### PR TITLE
8261142: AArch64: Incorrect instruction encoding when right-shifting vectors with shift amount equals to the element width

### PR DESCRIPTION
**Pull request by dgbo**
In vectorAPI, when right-shifting a vector with a shift equals to the element width, the shift is transformed to zero,
see `src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorOperators.java`:
```
    /** Produce {@code a>>>(n&(ESIZE*8-1))}. Integral only. */
    public static final /*bitwise*/ Binary LSHR = binary("LSHR", ">>>", VectorSupport.VECTOR_OP_URSHIFT, VO_SHIFT);
```

The aarch64 assembler generates wrong or illegal instructions in this case, e.g. for the JAVA code below on aarch64,
assembler call `__ ushr(dst, __ T8B, src, 0)`, the instruction generated is not `ushr dst.8B, src.8B, 0`, but `ushr dst.4H, src.4H, 16` instead.
According to local tests, JVM gives wrong results for byte/short and crashes with SIGILL for integer/long.
```
ByteVector vba = ByteVector.fromArray(byte64SPECIES, bytesA, 8 * i);
vbb.lanewise(VectorOperators.ASHR, 8).intoArray(arrBytes, 8 * i);
```

The legal right shift amount should be in the range 1 to the element width in bits on aarch64:
https://developer.arm.com/documentation/dui0801/f/A64-SIMD-Vector-Instructions/USHR--vector-?lang=en

This fix handles zero shift separately. If the shift is zero, it generates `orr` for right shift, `addv` for right shift and accumulate.
Verified with linux-aarch64-server-fastdebug, tier1. Also created a jtreg to reproduce the issue and for regression tests.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261142](https://bugs.openjdk.java.net/browse/JDK-8261142): AArch64: Incorrect instruction encoding when right-shifting vectors with shift amount equals to the element width

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2472/head:pull/2472`
`$ git checkout pull/2472`

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 1]
```diff
@@ -5277,9 +5282,14 @@ instruct vsra16B_imm(vecX dst, vecX src, immI shift) %{
   format %{ "sshr    $dst, $src, $shift\t# vector (16B)" %}
   ins_encode %{
     int sh = (int)$shift$$constant;
-    if (sh >= 8) sh = 7;
-    __ sshr(as_FloatRegister($dst$$reg), __ T16B,
-           as_FloatRegister($src$$reg), sh);
+    if (sh == 0) {
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5285]: If src and dst are the same reg, no need to emit code. Or maybe c2 can even be improved to optimize this out?

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 2]
```diff
@@ -5263,9 +5263,14 @@ instruct vsra8B_imm(vecD dst, vecD src, immI shift) %{
...
     int sh = (int)$shift$$constant;
-    if (sh >= 8) sh = 7;
-    __ sshr(as_FloatRegister($dst$$reg), __ T8B,
-           as_FloatRegister($src$$reg), sh);
+    if (sh == 0) {
+      __ orr(as_FloatRegister($dst$$reg), __ T8B,
+             as_FloatRegister($src$$reg), as_FloatRegister($src$$reg));
+    } else {
+      if (sh >= 8) sh = 7;
+      __ sshr(as_FloatRegister($dst$$reg), __ T8B,
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5271]: I think we should add an assert to make sure 0 is not passed to the assembler.

**Review by nsjian — COMMENTED** [review]
Thanks for the fix.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 1 above)
Comment (dgbo) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5285]: > If src and dst are the same reg, no need to emit code. 
If we want to do this enhancement, I think we need do it for left shifting and all SVE left/right shifting as well for completeness.
> Or maybe c2 can even be improved to optimize this (sh=0 case) out?
We can add code in `Ideal` to optimize it to ORR, but I'm not sure `orr` performs better than `shift` on other platforms.
Seems we have to created a generic new node to do `vector move` here.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 2 above)
Comment (dgbo) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5271]: Agree, I'll do this.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 [hunk 3]
```diff
@@ -2048,18 +2048,21 @@ instruct vsra$3$4_imm`'(vec$6 dst, vec$6 src, immI shift) %{
...
-    __ $2(as_FloatRegister($dst$$reg), __ T$3$5,
-           as_FloatRegister($src$$reg), sh);', `
-    __ $2(as_FloatRegister($dst$$reg), __ T$3$5,
-            as_FloatRegister($src$$reg),
-            (int)$shift$$constant);')
+    if (sh == 0) {
+      __ orr(as_FloatRegister($dst$$reg), __ ifelse($6, D, T8B, T16B),
+             as_FloatRegister($src$$reg), as_FloatRegister($src$$reg));
+    } else {ifelse($4, B,`
+      if (sh >= 8) sh = 7;
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2057]: I think it would be possible to move some of this logic from the AD file into MacroAssembler, with macros to generate the appropriate instruction based on their arguments. This might be cleaner: the logic here is very hard to follow.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 1 above)
Comment (nsjian) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5285]: I think with proper optimization, no move is required. But I agree it's beyond the scope of this patch. I will have a look.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 (hunk 3 above)
Comment (dgbo) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: I backed out the modifications of `aarch64_neon.ad` and `aarch64_neon_ad.m4`.
The `shift == 0` case is handled by the assembler now. Verified with the regression tests.

Comment (dgbo) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: > I think it would be possible to move some of this logic from the AD file into MacroAssembler, with macros to generate the appropriate instruction based on their arguments. This might be cleaner: the logic here is very hard to follow.

Hi, I moved the logic to the assembler.
The assembler will generate different instructions based on the value of `shift`.
If `shift == 0` and need not to accumulte, generated a `mov`.
If `shift == 0` and need to accumulte, generated an `add`.

Also added tests in `aarch64-asmtest.py` to verify the assembler modifications.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 4]
```diff
@@ -2684,12 +2684,28 @@ void mvnw(Register Rd, Register Rm,
...
+        f(0, 31), f(T & 1, 30), f(0b001110101, 29, 21);                 \
+        rf(Vn, 16), f(0b000111, 15, 10), rf(Vn, 5), rf(Vd, 0);          \
+      }                                                                 \
+    } else {                                                            \
+      int cVal = (1 << (((T >> 1) + 3) + (isSHR ? 1 : 0)));             \
+      int encodedShift = isSHR ? cVal - shift : cVal + shift;           \
+      f(0, 31), f(T & 1, 30), f(opc, 29), f(0b011110, 28, 23),          \
+      f(encodedShift, 22, 16); f(opc2, 15, 10), rf(Vn, 5), rf(Vd, 0);   \
+    }                                                                   \
   }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:2709]: Is this correct, according to the definition in the Architecture Reference Manual? It doesn't look like it to me. Assembler methods should generate bit patterns exactly as defined in the Manual. This logic should be in a MacroAssembler method.

Comment (dgbo) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp]: Hi, I moved the logic into MacroAssembler.
The assert is kept to make sure that we would never pass a zero right shift to assemlber.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 5]
```diff
@@ -527,6 +527,32 @@ class MacroAssembler: public Assembler {
...
+  void INSN(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, int shift) {  \
+    if (shift == 0) {                                                             \
+      Assembler::addv(Vd, T, Vd, Vn);                                             \
+    } else {                                                                      \
+      Assembler::INSN(Vd, T, Vn, shift);                                          \
+    }                                                                             \
+  }                                                                               \
+
+  WRAP(usra) WRAP(ssra)
+#undef WRAP
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:554]: Are ssra and usra tested by anything? I don't seem them accessed in the test case.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp [hunk 6]
```diff
@@ -527,6 +527,32 @@ class MacroAssembler: public Assembler {
     orr(Vd, T, Vn, Vn);
   }
 
+  // NEON shift instructions
+#define WRAP(INSN)                                                                \
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:531]: This comment should  be

```
  // AdvSIMD shift by immediate.     
  // These are "user friendly" variants which allow a shift count of 0.
```

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp (hunk 5 above)
Comment (dgbo) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp:555]: Updated. The `ssra/usra` are accessed by tests in `TestVectorShiftImmAndAccumulate.java`.
Manually injected error by changing `addv` to `subv` if shifting right and accumulating with 0, the tests failed as expected.

The `vba.add(vbb.lanewise(SHIFT, Imm))` pattern in `TestVectorShiftImmAndAccumulate.java` are actually the same with the original code in `TestVectorShiftImm.java`.
As of now, I have no idea why `ssra/usra` are not accessed by the previous test code.
The `vba.add(vbb.lanewise(SHIFT, Imm))` pattern should match `ssra/usra` anyway.
I think we need a separate investigation.

**Comment by dgbo** [comment]
Hi, @theRealAph 

I've rebased the tests so that sshr/ushr/sshl/ssra/usra are accessed in one jtreg `test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java`.
Local tests by manually injected error shows all instructions are covered by the jtreg case. Suggestions?

**Comment by theRealAph** [comment]
> Hi, @theRealAph
> 
> I've rebased the tests so that sshr/ushr/sshl/ssra/usra are accessed in one jtreg `test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java`.
> Local tests by manually injected error shows all instructions are covered by the jtreg case. Suggestions?

I'm not seeing sra used anywhere.

The problem I see with the tests is that the methods are large. This causes C2 to do a lot of spilling. Also, because the resuling code is intertwined and complex, it's very hard to debug.

It would be far better to do something like this:
```
    void long_shift_add(long arrLongs[][], LongVector vba, LongVector vbb, int i) {
        vba.add(vbb.lanewise(VectorOperators.LSHR, 37)).intoArray(arrLongs[op], 2 * i);
        vba.add(vbb.lanewise(VectorOperators.LSHR, 64)).intoArray(arrLongs[op  + 1], 2 * i);
        vba.add(vbb.lanewise(VectorOperators.LSHR, 99)).intoArray(arrLongs[op  + 2], 2 * i);
        vba.add(vbb.lanewise(VectorOperators.LSHR, 128)).intoArray(arrLongs[op + 3], 2 * i);
        vba.add(vbb.lanewise(VectorOperators.LSHR, 157)).intoArray(arrLongs[op + 4], 2 * i);
        vba.add(vbb.lanewise(VectorOperators.LSHR, 192)).intoArray(arrLongs[op + 5], 2 * i);
    }
```

**Comment by dgbo** [comment]
> > Local tests by manually injected error shows all instructions are covered by the jtreg case. Suggestions?
> 
> I'm not seeing `sra` used anywhere.
> 
> The problem I see with the tests is that the methods are large. This causes C2 to do a lot of spilling. Also, because the resuling code is intertwined and complex, it's very hard to debug.
> 
> It would be far better to do something like this:
> 
> ```
>     void long_shift_add(long arrLongs[][], LongVector vba, LongVector vbb, int i) {
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 37)).intoArray(arrLongs[op], 2 * i);
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 64)).intoArray(arrLongs[op  + 1], 2 * i);
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 99)).intoArray(arrLongs[op  + 2], 2 * i);
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 128)).intoArray(arrLongs[op + 3], 2 * i);
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 157)).intoArray(arrLongs[op + 4], 2 * i);
>         vba.add(vbb.lanewise(VectorOperators.LSHR, 192)).intoArray(arrLongs[op + 5], 2 * i);
>     }
> ```


Weird, I took a look at the the assembly, `ssra` did accessed by the tests on our server:
```
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:CompileThreshold=1000 -XX:CompileCommand=print,compiler/vectorapi/TestVectorShiftImm.shift_* test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java 64 &> assembly_vlen64.txt
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:CompileThreshold=1000 -XX:CompileCommand=print,compiler/vectorapi/TestVectorShiftImm.shift_* test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java 128 &> assembly_vlen128.txt
$ cat assembly_vlen*.txt | grep "ssra"
02c0    ssra    V18, V17, #37   # vector (2D)
02c8    ssra    V19, V17, #0    # vector (2D)
02d0    ssra    V20, V17, #35   # vector (2D)
0308    ssra    V18, V17, #29   # vector (2D)
0644    ssra    V18, V17, #37   # vector (2D)
064c    ssra    V19, V17, #0    # vector (2D)
0654    ssra    V20, V17, #35   # vector (2D)
0674    ssra    V18, V17, #29   # vector (2D)
0798    ssra    V18, V17, #37   # vector (2D)
07a0    ssra    V19, V17, #0    # vector (2D)
07a8    ssra    V20, V17, #35   # vector (2D)
07e0    ssra    V18, V17, #29   # vector (2D)
  0x0000ffff83f7e500:   ssra    v18.2d, v17.2d, #37         ;*aload_0 {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f7e510:   ssra    v20.2d, v17.2d, #35         ;*iand {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f7e548:   ssra    v18.2d, v17.2d, #29         ;*if_icmpne {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f7e884:   ssra    v18.2d, v17.2d, #37
  0x0000ffff83f7e894:   ssra    v20.2d, v17.2d, #35
  0x0000ffff83f7e8b4:   ssra    v18.2d, v17.2d, #29
  0x0000ffff83f7e9d8:   ssra    v18.2d, v17.2d, #37         ;*invokestatic broadcastInt {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f7e9e8:   ssra    v20.2d, v17.2d, #35         ;*invokestatic broadcastInt {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f7ea20:   ssra    v18.2d, v17.2d, #29         ;*checkcast {reexecute=0 rethrow=0 return_oop=0}
284     ssra    V18, V17, #9    # vector (4S)
28c     ssra    V19, V17, #0    # vector (4S)
294     ssra    V20, V17, #15   # vector (4S)
  0x0000ffff83f822c4:   ssra    v18.4s, v17.4s, #9          ;*invokedynamic {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff83f822d4:   ssra    v20.4s, v17.4s, #15
284     ssra    V18, V17, #1    # vector (8H)
28c     ssra    V19, V17, #8    # vector (8H)
...

```
Also injected error to `sshr+add` by:
```
--- a/src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp
+++ b/src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp
@@ -545,7 +545,7 @@ public:
 #define WRAP(INSN)                                                                \
   void INSN(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, int shift) {  \
     if (shift == 0) {                                                             \
-      Assembler::addv(Vd, T, Vd, Vn);                                             \
+      Assembler::subv(Vd, T, Vd, Vn);                                             \
     } else {                                                                      \
       Assembler::INSN(Vd, T, Vn, shift);                                          \
     }                                                                             \
```
The `shift+add` tests failed as expected:
```
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:CompileThreshold=1000 -XX:-TieredCompilation test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java 64
WARNING: Using incubator modules: jdk.incubator.vector
warning: using incubating module(s): jdk.incubator.vector
1 warning
Exception in thread "main" java.lang.RuntimeException: Test Failed, failed tests:
type SHORT index 19, operation ASHR_AND_ACCUMULATE, vector length 64.
type SHORT index 21, operation ASHR_AND_ACCUMULATE, vector length 64.
type SHORT index 23, operation ASHR_AND_ACCUMULATE, vector length 64.
type SHORT index 25, operation LSHR_AND_ACCUMULATE, vector length 64.
type SHORT index 27, operation LSHR_AND_ACCUMULATE, vector length 64.
type SHORT index 29, operation LSHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 19, operation ASHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 21, operation ASHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 23, operation ASHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 25, operation LSHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 27, operation LSHR_AND_ACCUMULATE, vector length 64.
type INTEGER index 29, operation LSHR_AND_ACCUMULATE, vector length 64.
...
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:CompileThreshold=1000 -XX:-TieredCompilation test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java 128
WARNING: Using incubator modules: jdk.incubator.vector
warning: using incubating module(s): jdk.incubator.vector
1 warning
Exception in thread "main" java.lang.RuntimeException: Test Failed, failed tests:
type LONG index 49, operation ASHR_AND_ACCUMULATE, vector length 128.
type LONG index 51, operation ASHR_AND_ACCUMULATE, vector length 128.
type LONG index 53, operation ASHR_AND_ACCUMULATE, vector length 128.
type LONG index 55, operation LSHR_AND_ACCUMULATE, vector length 128.
type LONG index 57, operation LSHR_AND_ACCUMULATE, vector length 128.
type LONG index 59, operation LSHR_AND_ACCUMULATE, vector length 128.
type SHORT index 49, operation ASHR_AND_ACCUMULATE, vector length 128.
type SHORT index 51, operation ASHR_AND_ACCUMULATE, vector length 128.
type SHORT index 53, operation ASHR_AND_ACCUMULATE, vector length 128.
...

```
Anyway, I extracted operations you suggested into `shift_op_*` methods.
Performed the error-injected experiments with the new tests on Kunpeng916 and re-checked the assembly output, results looks good.

The test command I used to run the newest tests are:
```
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:-TieredCompilation -XX:CompileThreshold=1000 -Dvlen=64 -XX:CompileCommand=print,compiler/vectorapi/TestVectorShiftImm.shift_* test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java &> assembly_vlen64.txt
$ ./build/linux-aarch64-server-fastdebug/images/jdk/bin/java --add-modules jdk.incubator.vector -XX:-TieredCompilation -XX:CompileThreshold=1000 -Dvlen=128 -XX:CompileCommand=print,compiler/vectorapi/TestVectorShiftImm.shift_* test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java &> assembly_vlen128.txt
$ cat assembly_vlen64.txt | grep ssra; cat assembly_vlen128.txt | grep ssra
```

**Comment by dgbo** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@openjdk.java.net):_
> 
> On 24/02/2021 07:33, Dong Bo wrote:
> 
> > Weird, I took a look at the the assembly, `ssra` did accessed by the tests on our server:
> 
> I don't doubt it, but the test code is so very complex that it can
> fall foul of heuristics given slightly changed circumstances. That's
> why good test cases are as simple as possible, and allow no room for
> variations because they do only one thing. Precise targeting should
> be the goal of HotSpot back-end test cases.
>
 
Understood, thanks. :-)
Does the newest version address the considerations?
I extracted the `shift`/`shift+add` operations into separate methods mostly as suggested in previous comments, something like:
```
    static int shift_op_long_ASHR_and_ADD(LongVector vba, LongVector vbb, long arrLongs[][], int end, int ind) {
        vba.add(vbb.lanewise(VectorOperators.ASHR, 37)).intoArray(arrLongs[end++], ind);
        vba.add(vbb.lanewise(VectorOperators.ASHR, 64)).intoArray(arrLongs[end++], ind);
        vba.add(vbb.lanewise(VectorOperators.ASHR, 99)).intoArray(arrLongs[end++], ind);
        vba.add(vbb.lanewise(VectorOperators.ASHR, 128)).intoArray(arrLongs[end++], ind);
        vba.add(vbb.lanewise(VectorOperators.ASHR, 157)).intoArray(arrLongs[end++], ind);
        vba.add(vbb.lanewise(VectorOperators.ASHR, 192)).intoArray(arrLongs[end++], ind);
        return end;
    }
```

**Comment by dgbo** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@openjdk.java.net):_
> 
> 
> I don't doubt it, but the test code is so very complex that it can
> fall foul of heuristics given slightly changed circumstances. That's
> why good test cases are as simple as possible, and allow no room for
> variations because they do only one thing. Precise targeting should
> be the goal of HotSpot back-end test cases.
> 

Test updated, all the operations to test are put in overloaded functions, `shift_with_op` and `shift_with_op_and_add`, repeatly called and tested by `shift` and `shift_and_accumulate` repectively with a loop.
Commands below are used to verify that `ssra` is accessed:
```
$ cp hsdis-aarch64.so ./build/linux-aarch64-server-fastdebug/images/jdk/lib/
$ jtreg -verbose:all -J-Djavatest.maxOutputSize=50000000 test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java | grep ssra
```

The tests do use `ssra` on our two different platforms, Kunpeng916 and Kunpeng920:
```
2b4 +   ssra    V16, V20, #17   # vector (2S)
2bc +   ssra    V17, V20, #0    # vector (2S)
2c4 +   ssra    V18, V20, #21   # vector (2S)
310 +   ssra    V16, V20, #12   # vector (2S)
  0x0000ffff9894e0f4:   ssra    v16.2s, v20.2s, #17         ;*invokespecial fromArray0Template {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff9894e104:   ssra    v18.2s, v20.2s, #21         ;*invokestatic arrayAddress {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff9894e150:   ssra    v16.2s, v20.2s, #12         ;*checkcast {reexecute=0 rethrow=0 return_oop=0}
2b0 +   ssra    V16, V20, #9    # vector (4H)
2b8 +   ssra    V17, V20, #0    # vector (4H)
2c0 +   ssra    V18, V20, #11   # vector (4H)
  0x0000ffff9894ae70:   ssra    v16.4h, v20.4h, #9          ;*invokevirtual vspecies {reexecute=0 rethrow=0 return_oop=0}
  0x0000ffff9894ae80:   ssra    v18.4h, v20.4h, #11         ;*invokestatic arrayAddress {reexecute=0 rethrow=0 return_oop=0}
: #     out( N1802 ) <- 298in( R29,  + R25B70, ssra    V16 ) #7
 + a4c     ssra         V17 + , spill V20, #0   # vector (8B)
        mov ssra    a4cV18R1 + , spill V20R11 -> [sp, #12],     # spill size = 32,
ssra    V16, V20, d08#3         # vector (8B)B150
...
```

Any comments?
Thanks.

File: test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+        }
+    }
+
+    /**
+     * Tests for type byte.
+     */
+
+    static int shift_with_op(VectorOperators.Binary op, ByteVector vbb,
+                             byte arr[][], int end, int ind) {
+        vbb.lanewise(op, 1).intoArray(arr[end++], ind);
```
Comment (theRealELiu) [inline test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java:133]: How about adding case 0 in this test? Those code was changed.

Comment (dgbo) [reply test/hotspot/jtreg/compiler/vectorapi/TestVectorShiftImm.java:134]: Thank you for watching this.
It can be covered by `lanewise(op, 8)`, `lanewise(op, 16)` and `lanewise(op, 24)`.
But I still added it, it looks clearer.

@theRealAph could you please take a look at the newest version?
The tests work well on our two very different aarch64 platforms.
Or the tests still looks too complicated?
Please let me know if you have any suggestions. Thanks.

**Review by theRealAph — APPROVED** [review]

**Comment by dgbo** [comment]
@theRealAph Thanks you for the review.

@nsjian @theRealELiu Hi, are you also ok with newest version? Thanks.

**Comment by theRealELiu** [comment]
> @theRealAph Thanks for the review.
> 
> @nsjian @theRealELiu Hi, are you also ok with newest version? Thanks.

Thanks for your fix! Looks good to me.

**Review by nsjian — APPROVED** [review]
